packages/ses/src/commons.js

```diff
@@ -378,3 +378,36 @@ export const FERAL_STACK_GETTER = feralStackGetter;
  * @type {((newValue: any) => void) | undefined}
  */
 export const FERAL_STACK_SETTER = feralStackSetter;
+
+const getAsyncGeneratorFunctionInstance = () => {
+  // Test for async generator function syntax support.
+  try {
+    // Wrapping one in an new Function lets the `hermesc` binary file
+    // parse the Metro js bundle without SyntaxError, to generate the
+    // optimised Hermes bytecode bundle, when `gradlew` is called to
+    // assemble the release build APK for React Native prod Android apps.
+    // Delaying the error until runtime lets us customise lockdown behaviour.
+    return new FERAL_FUNCTION(
+      'return (async function* AsyncGeneratorFunctionInstance() {})',
+    )();
+  } catch (e) {
+    // Note: `Error.prototype.jsEngine` is only set by React Native runtime, not Hermes:
+    // https://github.com/facebook/react-native/blob/main/packages/react-native/ReactCommon/hermes/executor/HermesExecutorFactory.cpp#L224-L230
+    if (e.name === 'SyntaxError') {
+      // Swallows Hermes error `async generators are unsupported` at runtime.
+      // @ts-expect-error ts(2554) Expected 0 arguments, but got 1. It refers to the Web API Window object, but on Hermes we expect 1 argument.
+      // eslint-disable-next-line no-undef
+      print('SES: Skipping async generators, unsupported on Hermes');
+      // Note: `console` is not a JS built-in, so Hermes engine throws:
+      // Uncaught ReferenceError: Property 'console' doesn't exist
+      // See: https://github.com/facebook/hermes/issues/675
+      // However React Native provides a `console` implementation when setting up error handling:
+      // https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Core/InitializeCore.js
+      return undefined;
+    } else {
+      throw e;
+    }
+  }
+};
+export const AsyncGeneratorFunctionInstance =
```

Comment:
Please add at least a comment or a type that makes clear the conditional nature of this export. Maybe even a rename to imply its conditional nature?

Since this export is exported static state that isn't entered as an intrinsic and so not hardened by lockdown, should we freeze it here before exporting? Is there any other non `FERAL_...` values exported by commons.js but untouched by lockdown that should be frozen before export?

Reply:
Just checking: We do not have any such problem with async functions or generator functions? Only with async generator functions?

Reply:
> Just checking: We do not have any such problem with async functions or generator functions? Only with async generator functions?

yes and we have a problemo with async arrow functions, it's abit of a mess, so i'll clarify:

generators are [supported](https://github.com/facebook/hermes/blob/main/doc/Features.md?plain=1#L21) on (correct docs)

async functions are [supported](https://github.com/facebook/hermes/issues/715#issuecomment-1083236894) (comment)
despite the error in the [codebase](https://github.com/facebook/hermes/blob/main/lib/IRGen/ESTreeIRGen-func.cpp#L213-L217)
despite listed in [docs](https://github.com/facebook/hermes/blob/main/doc/Features.md?plain=1#L34) as _In Progress_

it's actually async arrow functions aren't supported (not in docs/code) so i filed the [issue](https://github.com/facebook/hermes/issues/1395)
([fixed](https://github.com/facebook/hermes/issues/1395#issuecomment-2344662550) since only in Static Hermes)
so we're handling these on Hermes in this PR [packages/ses/scripts/hermes-transforms.js](https://github.com/endojs/endo/pull/2334/files#diff-119a8764a1d381d9018ad53b66952913680a53afd9c89f55955ca3dc8a0019a1)
for a Hermes flavoured SES shim

and then finally yes [async generators are unsupported](https://github.com/facebook/hermes/blob/main/lib/AST/SemanticValidator.cpp#L798-L800) is accurate (not mentioned in docs)
so we're handling that here with `AsyncGeneratorFunctionInstance` delaying the error till runtime
then hardening Hermes accordingly